ivadomed/testing.py

```diff
@@ -187,28 +196,55 @@ def run_inference(test_loader, model, model_params, testing_params, ofolder, cud
-                                                 fname_pred.split(".nii.gz")[0] + '_color.nii.gz',
-                                                 imed_utils.AXIS_DCT[testing_params['slice_axis']])
-
-    return np.concatenate(preds_npy_list, axis=0), np.concatenate(gt_npy_list, axis=0)
+                h_min, h_max, w_min, w_max, d_min, d_max = batch['input_metadata'][smp_idx][0]['coord']
+                num_pred = preds_cpu[smp_idx].shape[0]
+
+                first_sample_bool = not any([h_min, w_min, d_min])
+                if first_sample_bool:
+                    h, w, d = batch['input_metadata'][smp_idx][0]['index_shape']
+                    volume = torch.zeros((num_pred, h, w, d))
+                    weight_matrix = torch.zeros((num_pred, h, w, d))
+
+                last_sample_bool = h_max == h and w_max == w and d_max == d
+
+                # Average predictions
+                volume[:, h_min:h_max, w_min:w_max, d_min:d_max] += preds_cpu[smp_idx]
+                weight_matrix[:, h_min:h_max, w_min:w_max, d_min:d_max] += 1
+                if last_sample_bool:
+                    volume /= weight_matrix
+
+                pred_undo, metadata = testing_params["undo_transforms"](volume,
+                                                                        batch['gt_metadata'][smp_idx],
+                                                                        data_type='gt')
+                fname_ref = metadata[0]['gt_filenames'][0]
+                # Indicator of last batch
+                if last_sample_bool:
+                    pred_undo = np.array(pred_undo)
+                    fname_pred = os.path.join(ofolder, fname_ref.split('/')[-1])
+                    fname_pred = fname_pred.split(testing_params['target_suffix'][0])[0] + '_pred.nii.gz'
+                    # If uncertainty running, then we save each simulation result
+                    if testing_params['uncertainty']['applied']:
+                        fname_pred = fname_pred.split('.nii.gz')[0] + '_' + str(i_monteCarlo).zfill(2) + '.nii.gz'
+
+                    # Choose only one modality
+                    output_nii = imed_utils.pred_to_nib(data_lst=[pred_undo],
+                                                        z_lst=[],
+                                                        fname_ref=fname_ref,
+                                                        fname_out=fname_pred,
+                                                        slice_axis=imed_utils.AXIS_DCT[testing_params['slice_axis']],
+                                                        kernel_dim='3d',
+                                                        bin_thr=0.5 if testing_params["binarize_prediction"] else -1)
+                    preds_npy_list.append(output_nii.get_fdata().transpose(3, 0, 1, 2))
+                    gt_lst = []
+                    for gt in metadata[0]['gt_filenames']:
```

Comment:
Here there is a mismatch between fname_pred and fname_gt:
`metadata[0]['gt_filenames']` refers to the next image, which is currently being filled.